pkg/release/release.go

```diff
@@ -1,556 +1,342 @@
+	if c.ChartCache == "" {
+		c.ChartCache = "/tmp"
+	}
+	return c
 }
 
-type InstallOptions struct {
-	DryRun    bool
-	ReuseName bool
+// Release holds the elements required to perform a Helm release,
+// and provides the methods to perform a sync or uninstall.
+type Release struct {
+	logger             log.Logger
+	coreV1Client       corev1client.CoreV1Interface
+	helmReleaseClient  v1client.HelmV1Interface
+	gitChartSourceSync *chartsync.GitChartSourceSync
+	config             Config
 }
 
-// New creates a new Release instance.
-func New(logger log.Logger, helmClient helm.Client) *Release {
+func New(logger log.Logger, coreV1Client corev1client.CoreV1Interface, helmReleaseClient v1client.HelmV1Interface,
+	gitChartSourceSync *chartsync.GitChartSourceSync, config Config) *Release {
+
 	r := &Release{
-		logger:     logger,
-		helmClient: helmClient,
+		logger:             logger,
+		coreV1Client:       coreV1Client,
+		helmReleaseClient:  helmReleaseClient,
+		gitChartSourceSync: gitChartSourceSync,
+		config:             config.WithDefaults(),
 	}
 	return r
 }
 
-// GetUpgradableRelease returns a release if the current state of it
-// allows an upgrade, a descriptive error if it is not allowed, or
-// nil if the release does not exist.
-func (r *Release) GetUpgradableRelease(namespace string, releaseName string) (*helm.Release, error) {
-	rel, err := r.helmClient.Status(releaseName, helm.StatusOptions{Namespace: namespace})
-	if err != nil {
-		if strings.Contains(err.Error(), "not found") {
-			return nil, nil
+// Sync synchronizes the given `v1.HelmRelease` with Helm.
+func (r *Release) Sync(client helm.Client, hr *v1.HelmRelease) (*v1.HelmRelease, error) {
+	defer status.SetObservedGeneration(r.helmReleaseClient.HelmReleases(hr.Namespace), *hr, hr.Generation)
+
+	logger := releaseLogger(r.logger, hr)
+
+	// Ensure we have the chart for the release, construct the path
+	// to the chart and get the revision.
+	var chartPath, revision string
+	switch {
+	case hr.Spec.GitChartSource != nil:
+		s, _ := r.gitChartSourceSync.Load(hr)
+		if s == nil {
+			// Chart source has been requested, but is not ready yet,
+			// wait for the next signal...
+			return hr, nil
 		}
-		return nil, err
-	}
-	if rel.Info == nil {
-		return nil, fmt.Errorf("release exists but can not determine if it can be upgraded due to missing metadata")
-	}
-	switch rel.Info.Status {
-	case helm.StatusDeployed:
-		return &rel, nil
-	case helm.StatusFailed:
-		return nil, fmt.Errorf("release requires a rollback before it can be upgraded (%s)", rel.Info.Status)
-	case helm.StatusPendingInstall,
-		helm.StatusPendingUpgrade,
-		helm.StatusPendingRollback:
-		return nil, fmt.Errorf("operation pending for release (%s)", rel.Info.Status)
-	default:
-		return nil, fmt.Errorf("current state prevents it from being upgraded (%s)", rel.Info.Status)
-	}
-}
+		// Lock the resource so that it does not get updated by
+		// the chart sync process.
+		s.Lock()
+		defer s.Unlock()
```

Comment:
Do you need to hold the lock throughout the execution of the rest of function?

Reply:
Yes, as otherwise a signal could replace the working clone and clean up the old one while the release has not happened yet.